worker/src/primary_connector.rs

```diff
@@ -3,94 +3,135 @@
 // SPDX-License-Identifier: Apache-2.0
 use config::Committee;
 use crypto::traits::VerifyingKey;
+use futures::FutureExt;
 use multiaddr::Multiaddr;
+use network::{BoundedExecutor, CancelHandler, RetryConfig, MAX_TASK_CONCURRENCY};
 use primary::WorkerPrimaryMessage;
 use tokio::{
+    runtime::Handle,
     sync::{mpsc::Receiver, watch},
     task::JoinHandle,
 };
 use tonic::transport::Channel;
-use types::{BincodeEncodedPayload, Reconfigure, WorkerToPrimaryClient};
+use types::{BincodeEncodedPayload, ReconfigureNotification, WorkerToPrimaryClient};
 
 // Send batches' digests to the primary.
 pub struct PrimaryConnector<PublicKey: VerifyingKey> {
+    /// The public key of this authority.
+    name: PublicKey,
+    /// The committee information.
+    committee: Committee<PublicKey>,
     /// Receive reconfiguration updates.
-    rx_reconfigure: watch::Receiver<Reconfigure<PublicKey>>,
+    rx_reconfigure: watch::Receiver<ReconfigureNotification<PublicKey>>,
     /// Input channel to receive the messages to send to the primary.
-    rx_digest: Receiver<WorkerPrimaryMessage>,
+    rx_digest: Receiver<WorkerPrimaryMessage<PublicKey>>,
     /// A network sender to send the batches' digests to the primary.
-    primary_client: PrimaryClient,
+    primary_client: WorkerToPrimaryNetwork,
 }
 
 impl<PublicKey: VerifyingKey> PrimaryConnector<PublicKey> {
     pub fn spawn(
         name: PublicKey,
         committee: Committee<PublicKey>,
-        rx_reconfigure: watch::Receiver<Reconfigure<PublicKey>>,
-        rx_digest: Receiver<WorkerPrimaryMessage>,
+        rx_reconfigure: watch::Receiver<ReconfigureNotification<PublicKey>>,
+        rx_digest: Receiver<WorkerPrimaryMessage<PublicKey>>,
     ) -> JoinHandle<()> {
-        let address = committee
-            .primary(&name)
-            .expect("Our public key is not in the committee")
-            .worker_to_primary;
         tokio::spawn(async move {
             Self {
+                name,
+                committee,
                 rx_reconfigure,
                 rx_digest,
-                primary_client: PrimaryClient::new(address),
+                primary_client: WorkerToPrimaryNetwork::default(),
             }
             .run()
             .await;
         })
     }
 
     async fn run(&mut self) {
+        // Ensure we don't drop the handle too fast.
+        let mut _handle: CancelHandler<()>;
+
         loop {
             tokio::select! {
                 // Send the digest through the network.
                 Some(digest) = self.rx_digest.recv() => {
-                    // We don't care about the error
-                    let _ = self.primary_client.send(&digest).await;
+                    let address = self.committee
+                        .primary(&self.name)
+                        .expect("Our public key is not in the committee")
+                        .worker_to_primary;
+                    _handle = self.primary_client.send(address, &digest).await;
```

Comment:
Very interest, when I do this:
```
 let handle = self.primary_client.send(address, &digest).await;
 handle.await;
```

the issues observed here https://github.com/MystenLabs/narwhal/pull/489#issuecomment-1192490074 are resolved and the tests are able to pass correctly. So we are probably looking to some sort of concurrency issue? Need to look further into it.

Reply:
Ok , so this is not a concurrency issue. The `CancelHandler` returned by the `send` method is aborting the task when dropped: https://github.com/MystenLabs/narwhal/blob/main/network/src/lib.rs#L29

So based on the info I posted here https://github.com/MystenLabs/narwhal/pull/489#issuecomment-1192490074 (about handlers getting overwritten) , indeed when the previous handler is dropped it has a consequence for the underlying task to get cancelled. That explains what we see here.

So we probably have two ways forward:
1. Don't use the `CancelHandle` struct, but rather return just a JoinHandle
2. Save the handles until they are finished - maybe we don't want to do that though, since the approach is fire and forget

Reply:
This makes sense, when I was investigating the failure I tried running the demo client against this PR and it would fail on calls to endpoints other than `GetCollections`. (i.e. `Rounds`, `ReadCausal`, etc.). Consistently seeing connections being dropped in the primaries.

```
2022-07-21T22:02:27.082935Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "error reading a body from connection: stream error received: stream no longer needed", source: Some(hyper::Error(Body, Error { kind: Reset(StreamId(1), CANCEL, Remote) })) }
```
[primary-1.log](https://github.com/MystenLabs/narwhal/files/9169689/primary-1.log)
[demo_client.log](https://github.com/MystenLabs/narwhal/files/9169694/demo_client.log)

Reply:
🫢 wow that was silly! thanks for your help, it was right under my nose. it probably means I need a vacation

Reply:
To be honest its not apparent the functionality of the `CancelHandler` (aborting the task when dropped) - I mean someone could easily miss that if not familiar with it (I wasn't as well). We'll probably rename this in a separate PR and add some comments as well so it becomes more apparent what someone should expect.

Reply:
I would point out the reason we wrap the `.broadcast` and `.send` operations on primaries and workers is that they are on an infinite retry from a network PoV -> if you're not baby-sitting the cancellation handlers, you should probably not use those operations, but rather their non-infinite-retry variants (`unreliable_broadcast`, `unreliable_send`).

Reply:
yes I agree, but it is probably a bad idea to rename all CancelHandler in this PR (it is already quite big)